### PR TITLE
E2E Conv2D Implementation

### PR DESCRIPTION
**Pull request by LPanosTT**
Changes:
- Updated MLIR to latest main and added my changes
- TTIR, TTNN MLIR support
- TTNN runtime support
- ToMemoryConfig ops should not be placed on the input activations, weight, or bias of the conv2d op as the conv2d op transforms these tensors on host and then moves them to device on its own (hack here)
- Typecast on the output ToMemoryConfig doesn't work on host (hack here)
-  MaxPool2d op definitions in TTIR and TTNN

TTForge PR: https://github.com/tenstorrent/tt-forge-fe/pull/168

### CODE REVIEW

File: include/ttmlir/Dialect/TTIR/IR/TTIROps.td [hunk 1]
```diff
@@ -340,6 +340,62 @@ def TTIR_TransposeOp : TTIR_DPSOp<"transpose"> {
...
+                         SI32Attr:$kernel_width,
+                         SI32Attr:$stride_height,
+                         SI32Attr:$stride_width,
+                         SI32Attr:$dilation,
+                         SI32Attr:$groups,
+                         SI32Attr:$padding_left,
+                         SI32Attr:$padding_right,
+                         SI32Attr:$padding_top,
+                         SI32Attr:$padding_bottom,
+                         SI32Attr:$is_convtranspose2d,
```
Comment (nvukobratTT) [inline include/ttmlir/Dialect/TTIR/IR/TTIROps.td:363]: Do we want to have a separated op named Conv2D Transpose?

File: include/ttmlir/Dialect/TTIR/IR/TTIROps.td [hunk 2]
```diff
@@ -340,6 +340,62 @@ def TTIR_TransposeOp : TTIR_DPSOp<"transpose"> {
     let hasVerifier = 1;
 }
 
+def TTIR_Conv2dOp : TTIR_DPSOp<"conv2d"> {
+    let summary = "Conv2d operation.";
+    let description = [{
+      Conv2d operation.
```
Comment (nvukobratTT) [inline include/ttmlir/Dialect/TTIR/IR/TTIROps.td:346]: This is more for MLIR owners to confirm, but it'll be good to focus on more detailed descriptions of the ops we're bringing up.

File: include/ttmlir/Dialect/TTIR/IR/TTIROps.td [hunk 3]
```diff
@@ -340,6 +340,62 @@ def TTIR_TransposeOp : TTIR_DPSOp<"transpose"> {
...
+    let results = (outs AnyRankedTensor:$result);
+
+    let extraClassDeclaration = [{
+      MutableOperandRange getDpsInitsMutable() { return getOutputMutable(); }
+    }];
+
+    let hasVerifier = 1;
+}
+
+def TTIR_ConvertConvWeightTensorToTiledLayoutOp : TTIR_DPSOp<"convert_conv_weight_tensor_to_tiled_layout"> {
```
Comment (nvukobratTT) [inline include/ttmlir/Dialect/TTIR/IR/TTIROps.td:378]: Is this more of a "pass" rather than an "op"?

File: lib/Dialect/TTIR/IR/TTIROps.cpp [hunk 4]
```diff
@@ -163,6 +163,42 @@ ::mlir::LogicalResult mlir::tt::ttir::MatmulOp::verify() {
 }
 // ANCHOR_END: adding_an_op_matmul_ttir_verify
 
+::mlir::LogicalResult mlir::tt::ttir::Conv2dOp::verify() {
```
Comment (nvukobratTT) [inline lib/Dialect/TTIR/IR/TTIROps.cpp:166]: Are we aware of more tt-nn-related constraints? E.g. wight layout, supported attributes, etc.

File: third_party/CMakeLists.txt [hunk 5]
```diff
@@ -42,7 +42,7 @@ ExternalProject_Add(
   PREFIX ${TTMLIR_SOURCE_DIR}/third_party/tt-metal
   CMAKE_GENERATOR Ninja
   CMAKE_ARGS
-    -DCMAKE_BUILD_TYPE=Release
+    -DCMAKE_BUILD_TYPE=Debug
```
Comment (nvukobratTT) [inline third_party/CMakeLists.txt:45]: Reminder to revert this change before merge

File: include/ttmlir/Dialect/TTIR/IR/TTIROps.td (hunk 1 above)
Comment (svuckovicTT) [reply include/ttmlir/Dialect/TTIR/IR/TTIROps.td:363]: Yeah, we definitely want to separate `Conv2d` and `ConvTranspose2d`

File: include/ttmlir/Dialect/TTIR/IR/TTIROps.td (hunk 3 above)
Comment (svuckovicTT) [reply include/ttmlir/Dialect/TTIR/IR/TTIROps.td:378]: Yes, we should move this. There's two places I think this might go:
1. `TTIRLayout` pass (`lib/Dialect/TTIR/Transforms/Passes.cpp`)
2. `TTIRToTTNN` dialect conversion (`lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp`)

I think `1` is a lot cleaner, we should do it there.

File: lib/Dialect/TTIR/Transforms/Passes.cpp [hunk 6]
```diff
@@ -565,9 +565,12 @@ class TTIRLayoutDPSOperandsRewriter
 
     assert(op->template hasTrait<TTIROp::Trait>());
     bool modified = false;
+    int idx = -1;
     for (auto &operand : op->getOpOperands()) {
+      idx++;
       bool isResult = op.isDpsInit(&operand);
-
+      if (idx == 0 or idx == 1 or idx == 2)
```
Comment (svuckovicTT) [inline lib/Dialect/TTIR/Transforms/Passes.cpp:572]: nit: `if (idx >=0 and idx <= 2)`

File: lib/Dialect/TTIR/Transforms/Passes.cpp [hunk 7]
```diff
@@ -565,9 +565,12 @@ class TTIRLayoutDPSOperandsRewriter
 
     assert(op->template hasTrait<TTIROp::Trait>());
     bool modified = false;
+    int idx = -1;
```
Comment (svuckovicTT) [inline lib/Dialect/TTIR/Transforms/Passes.cpp:568]: What's the rationale for skipping the first 3 operands here?

File: lib/Dialect/TTNN/IR/TTNNOps.cpp [hunk 8]
```diff
@@ -121,6 +121,42 @@ ::mlir::LogicalResult mlir::tt::ttnn::MatmulOp::verify() {
...
+  ::mlir::RankedTensorType inputType = getInput().getType();
+  ::mlir::RankedTensorType weightType = getWeight().getType();
+  ::mlir::RankedTensorType biasType = getBias().getType();
+  // ::mlir::RankedTensorType outputType = getOutput().getType();
+  auto inputShape = inputType.getShape();
+  auto weightShape = weightType.getShape();
+  auto biasShape = biasType.getShape();
+  // auto outputShape = outputType.getShape();
+  if (inputShape.size() < 3) {
+    return emitOpError("Input must be at least a 3D tensor");
```
Comment (svuckovicTT) [inline lib/Dialect/TTNN/IR/TTNNOps.cpp:134]: Not related to this change, but this makes me think that having a canonicalizer that makes sure there's batch size in ops' tensors would be useful. @nsmithtt thoughts?

File: lib/Dialect/TTNN/IR/TTNNOps.cpp [hunk 9]
```diff
@@ -121,6 +121,42 @@ ::mlir::LogicalResult mlir::tt::ttnn::MatmulOp::verify() {
...
+  }
+  if (weightShape.size() != 4) {
+    return emitOpError("Weight must be a 4D tensor");
+  }
+  if (biasShape.size() != 4) {
+    return emitOpError("Weight must be a 4D tensor");
```
Comment (svuckovicTT) [inline lib/Dialect/TTNN/IR/TTNNOps.cpp:140]: "Weight" -> "Bias" in error msg.

Shouldn't the bias be a 1d tensor?

File: lib/Dialect/TTNN/IR/TTNNOps.cpp [hunk 10]
```diff
@@ -121,6 +121,42 @@ ::mlir::LogicalResult mlir::tt::ttnn::MatmulOp::verify() {
...
+  ::mlir::RankedTensorType biasType = getBias().getType();
+  // ::mlir::RankedTensorType outputType = getOutput().getType();
+  auto inputShape = inputType.getShape();
+  auto weightShape = weightType.getShape();
+  auto biasShape = biasType.getShape();
+  // auto outputShape = outputType.getShape();
+  if (inputShape.size() < 3) {
+    return emitOpError("Input must be at least a 3D tensor");
+  }
+  if (weightShape.size() != 4) {
```
Comment (svuckovicTT) [inline lib/Dialect/TTNN/IR/TTNNOps.cpp:136]: nit: prefer to use `weightType.getRank()` vs `weightShape.size()`.

File: lib/Dialect/TTIR/Transforms/Passes.cpp (hunk 6 above)
Comment (rpavlovicTT) [reply lib/Dialect/TTIR/Transforms/Passes.cpp:572]: And please use C++ logical operators such as `&&, ||`.

Comment (svuckovicTT) [reply lib/Dialect/TTIR/Transforms/Passes.cpp:572]: Ah sorry, my bad, guess I'm too used to python now :)

`if (idx >=0 && idx <= 2)`

File: include/ttmlir/Dialect/TTIR/IR/TTIROps.td (hunk 3 above)
Comment (LPanosTT) [reply include/ttmlir/Dialect/TTIR/IR/TTIROps.td:378]: So I added this because it's an actual op in TTNN. What I didn't realize is that its also called within the TTNN Conv2D op, so I actually don't need it to bringup Conv2D.

File: lib/Dialect/TTIR/Transforms/Passes.cpp (hunk 7 above)
Comment (LPanosTT) [reply lib/Dialect/TTIR/Transforms/Passes.cpp:568]: This is one of the reasons why this PR isn't ready yet.

So the TTNN conv2d op itself moved the input, weights, and biases from host onto device. If those tensors are already on device when the op is called it causes errors. My understanding is that we want to stop moving op outputs on and off host between every op, so something will have to change about the TTNN Conv2D op.

File: lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp [hunk 11]
```diff
@@ -199,7 +260,9 @@ void populateTTIRToTTNNPatterns(MLIRContext *ctx, RewritePatternSet &patterns,
            EmbeddingOpConversionPattern,
            SoftmaxOpConversionPattern,
            TransposeOpConversionPattern,
-           MatmulOpConversionPattern
+           SoftmaxOpConversionPattern,
+           MatmulOpConversionPattern,
+           Conv2dOpConversionPattern
```
Comment (nsmithtt) [inline lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp:265]: Is there no maxpool conversion?

File: lib/Conversion/TTNNToEmitC/TTNNToEmitC.cpp [hunk 12]
```diff
@@ -200,6 +200,10 @@ void populateTTNNToEmitCPatterns(mlir::MLIRContext *ctx,
   patterns.add<DefaultOpConversionPattern<ttnn::SumOp>>(typeConverter, ctx);
   patterns.add<DefaultOpConversionPattern<ttnn::MeanOp>>(typeConverter, ctx);
 
+  // Conv ops
+  //
+  patterns.add<DefaultOpConversionPattern<ttnn::Conv2dOp>>(typeConverter, ctx);
```
Comment (nsmithtt) [inline lib/Conversion/TTNNToEmitC/TTNNToEmitC.cpp:205]: No maxpool conversion?

File: lib/Dialect/TTIR/Transforms/Passes.cpp [hunk 13]
```diff
@@ -568,6 +568,11 @@ class TTIRLayoutDPSOperandsRewriter
     for (auto &operand : op->getOpOperands()) {
       bool isResult = op.isDpsInit(&operand);
 
+      // @lpanos : TTNN Conv2d moves input, weight, and bias from host to device
```
Comment (nsmithtt) [inline lib/Dialect/TTIR/Transforms/Passes.cpp:571]: Can you remove your username?

I think it's also only true for weight and bias.

File: lib/Dialect/TTNN/IR/TTNNOps.cpp (hunk 8 above)
Comment (nsmithtt) [reply lib/Dialect/TTNN/IR/TTNNOps.cpp:152]: One of the goals of tensor layout is not to enforce a particular tensor rank.  If you mean specifically for the ttnn dialect though (not ttir) then that could be useful. I think they automatically snap to 4d in a lot of cases.

File: runtime/lib/ttnn/program.cpp [hunk 14]
```diff
@@ -400,6 +404,42 @@ run(::tt::target::ttnn::MatmulOp const *op, ::ttnn::Device &device,
...
+static void
+run(::tt::target::ttnn::Conv2dOp const *op, ::ttnn::Device &device,
+    std::unordered_map<std::uint32_t, ::ttnn::Tensor *> &liveTensors,
+    std::list<::ttnn::Tensor> &tensorPool) {
+  auto &input = *liveTensors.at(op->in0()->global_id());
+  auto &weight = *liveTensors.at(op->in1()->global_id());
+  auto config = ::ttnn::operations::conv2d::Conv2dConfig();
+  config.dtype = input.dtype();
+  config.weights_dtype = weight.dtype();
+  ::ttnn::Tensor ans;
```
Comment (nsmithtt) [inline runtime/lib/ttnn/program.cpp:416]: I think `out` is more in line with the naming in flatbuffer and other places in runtime.

File: runtime/lib/ttnn/program.cpp [hunk 15]
```diff
@@ -400,6 +404,42 @@ run(::tt::target::ttnn::MatmulOp const *op, ::ttnn::Device &device,
...
+  } else {
+    ans = std::get<0>(::ttnn::operations::conv2d::conv2d<::ttnn::Device>(
+        input, weight, &device, op->in_channels(), op->out_channels(),
+        op->batch_size(), op->input_height(), op->input_width(),
+        {op->kernel_height(), op->kernel_width()},
+        {op->stride_height(), op->stride_width()},
+        {op->padding_height(), op->padding_width()},
+        {op->dilation_height(), op->dilation_width()}, op->groups(),
+        std::nullopt, config));
+  }
```
Comment (nsmithtt) [inline runtime/lib/ttnn/program.cpp:436]: Can we rewrite this whole if/else?  Let's make the setting bias tensor conditional and then call the function in a single place:
```c++
std::optional<::ttnn::Tensor> bias = op->in2() ? *liveTensors.at(op->in2()->global_id()) : std::nullopt;
auto& out = std::get<0>(::ttnn::operations::conv2d::conv2d<::ttnn::Device>(..., bias, ....);
```

File: runtime/lib/ttnn/runtime.cpp [hunk 16]
```diff
@@ -54,7 +54,7 @@ Device openDevice(std::vector<int> const &deviceIds,
                   std::vector<std::uint8_t> const &numHWCQs) {
   assert(deviceIds.size() == 1 && "Only one device is supported for now");
   assert(numHWCQs.empty() && "HWCQs are not supported for now");
-  auto &device = ::ttnn::open_device(deviceIds.front());
+  auto &device = ::ttnn::open_device(deviceIds.front(), (1 << 15));
```
Comment (nsmithtt) [inline runtime/lib/ttnn/runtime.cpp:57]: Can we give the magic number a name?
```c++
constexpr std::size_t kL1SmallSize = 1 << 15;
```

File: test/ttmlir/Dialect/TTNN/simple_conv.mlir [hunk 17]
```diff
@@ -0,0 +1,10 @@
+// RUN: ttmlir-opt --ttir-load-system-desc --ttir-layout --ttnn-open-device --convert-ttir-to-ttnn %s | FileCheck %s
+#any_device_tile = #tt.operand_constraint<dram|l1|tile|any_device_tile>
+module attributes {} {
+  func.func @forward(%arg0: tensor<1x32x32x1xbf16>, %arg1: tensor<1x1x3x3xbf16>, %arg2: tensor<1x32x32x1xbf16>) -> tensor<1x32x32x1xbf16> {
+    %0 = tensor.empty() : tensor<1x32x32x1xbf16>
+    // CHECK: %[[C:.*]] = "ttnn.conv2d"[[C:.*]]
+    %1 = "ttir.conv2d"(%arg0, %arg1, %arg2, %0) <{stride_height=1: si32, stride_width=1: si32, dilation_height=1: si32, dilation_width=1: si32, groups=1: si32, padding_left=1: si32, padding_right=1: si32, padding_top=1: si32, padding_bottom=1: si32, is_convtranspose2d=0: si32, output_height_transpose=0: si32, output_width_transpose=0: si32, stride_transpose=0: si32, operand_constraints = [#any_device_tile, #any_device_tile, #any_device_tile, #any_device_tile]}> : (tensor<1x32x32x1xbf16>, tensor<1x1x3x3xbf16>, tensor<1x32x32x1xbf16>, tensor<1x32x32x1xbf16>) -> tensor<1x32x32x1xbf16>
```
Comment (nsmithtt) [inline test/ttmlir/Dialect/TTNN/simple_conv.mlir:7]: We need a test for maxpool too.  I'm ok with breaking maxpool into a separate change if you want.

File: lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp (hunk 11 above)
Comment (LPanosTT) [reply lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp:265]: Not yet, I'll bringup the maxpool op when tt-metal gets uplifted. Should I delete the op definitions from this branch?

File: lib/Conversion/TTNNToEmitC/TTNNToEmitC.cpp (hunk 12 above)
Comment (LPanosTT) [reply lib/Conversion/TTNNToEmitC/TTNNToEmitC.cpp:205]: Not yet, I'll bringup the maxpool op when tt-metal gets uplifted. Should I delete the op definitions from this branch?

File: lib/Dialect/TTIR/Transforms/Passes.cpp (hunk 13 above)
Comment (LPanosTT) [reply lib/Dialect/TTIR/Transforms/Passes.cpp:571]: So in `third_party/tt-mlir/third_party/tt-metal/src/tt-metal/ttnn/cpp/ttnn/operations/conv2d/conv2d.cpp` line 417 you'll see an if-else block that handles the input tensor differently whether or not it is on device. I've tried enabling the movement of the input tensor to device pre-conv2d, and I am seeing data mismatches on the output. I'm unsure why exactly this is. The TTNN test in `third_party/tt-mlir/third_party/tt-metal/src/tt-metal/tests/ttnn/unit_tests/operations/test_new_conv2d.py` looks like it does not move the tensors to device before executing `conv2d` either...

File: runtime/lib/ttnn/program.cpp (hunk 15 above)
Comment (LPanosTT) [reply runtime/lib/ttnn/program.cpp:436]: Will do

File: lib/Conversion/TTNNToEmitC/TTNNToEmitC.cpp (hunk 12 above)
Comment (nsmithtt) [reply lib/Conversion/TTNNToEmitC/TTNNToEmitC.cpp:205]: Yeah I think so, we want to get in the habit of having a unit test with every commit, with some exceptions.  But this is a case since the max pool isn't functional anyway, it'd be better to just wait until we can test it.

File: lib/Dialect/TTIR/Transforms/Passes.cpp (hunk 13 above)
Comment (LPanosTT) [reply lib/Dialect/TTIR/Transforms/Passes.cpp:571]: The memory config op placed on the output of the first conv moves it to host I believe.

File: runtime/lib/ttnn/program.cpp [hunk 18]
```diff
@@ -86,6 +89,7 @@ updateLayoutAndDataType(const ::ttnn::Tensor &inputTensor,
     outputTensor = ::untilize(outputTensor);
   }
   if (shouldConvertDataType) {
+    // (void)convertDataType;
```
Comment (nsmithtt) [inline runtime/lib/ttnn/program.cpp:92]: Stray comment, please remove before merging

**Review by nsmithtt — APPROVED** [review]

File: include/ttmlir/Dialect/TTNN/IR/TTNNOps.td [hunk 19]
```diff
@@ -239,6 +239,60 @@ def TTNN_MatmulOp : TTNN_NamedDPSOp<"matmul"> {
...
+    let results = (outs AnyRankedTensor:$result);
+
+    let extraClassDeclaration = [{
+      MutableOperandRange getDpsInitsMutable() { return getOutputMutable(); }
+    }];
+
+    let hasVerifier = 1;
+}
+
+def TTNN_ConvertConvWeightTensorToTiledLayoutOp : TTNN_NamedDPSOp<"convert_conv_weight_tensor_to_tiled_layout"> {
```
[the remark this answers is not in the document]
Comment (LPanosTT) [reply include/ttmlir/Dialect/TTNN/IR/TTNNOps.td]: removed this

File: lib/Dialect/TTIR/Transforms/Passes.cpp (hunk 6 above)
Comment (LPanosTT) [reply lib/Dialect/TTIR/Transforms/Passes.cpp]: Fixed this

File: lib/Conversion/TTNNToEmitC/TTNNToEmitC.cpp (hunk 12 above)
Comment (LPanosTT) [reply lib/Conversion/TTNNToEmitC/TTNNToEmitC.cpp:207]: Removed maxpool dead code

File: lib/Dialect/TTNN/IR/TTNNOps.cpp (hunk 9 above)
Comment (LPanosTT) [reply lib/Dialect/TTNN/IR/TTNNOps.cpp:259]: Added that check

File: include/ttmlir/Target/TTNN/program.fbs [hunk 20]
```diff
@@ -99,6 +99,27 @@ table MatmulOp {
 }
 // ANCHOR_END: adding_an_op_matmul_fbs
 
+table Conv2dOp {
+  in0: tt.target.TensorRef;
+  in1: tt.target.TensorRef;
+  in2: tt.target.TensorRef;
```
Comment (svuckovicTT) [inline include/ttmlir/Target/TTNN/program.fbs:105]: I would prefer it if we named these `input`, `weight`, `bias`, similiarly to embedding op above. It makes it easier to work with.

File: lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp [hunk 21]
```diff
@@ -12,6 +12,8 @@
 #include "mlir/Support/LogicalResult.h"
 #include "mlir/Transforms/DialectConversion.h"
 #include "llvm/Support/Casting.h"
+#include <cassert>
+#include <cstdint>
```
Comment (svuckovicTT) [inline lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp:16]: Check if we really need these and sort according to style guide: https://llvm.org/docs/CodingStandards.html#include-style

File: lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp [hunk 22]
```diff
@@ -204,6 +206,65 @@ class MatmulOpConversionPattern : public OpConversionPattern<ttir::MatmulOp> {
...
+    auto kernel_height =
+        rewriter.getI32IntegerAttr(kernel_shape[kernel_shape.size() - 2]);
+    auto kernel_width =
+        rewriter.getI32IntegerAttr(kernel_shape[kernel_shape.size() - 1]);
+
+    auto stride_height = rewriter.getI32IntegerAttr(adaptor.getStrideHeight());
+    auto stride_width = rewriter.getI32IntegerAttr(adaptor.getStrideWidth());
+
+    assert(adaptor.getPaddingBottom() == adaptor.getPaddingTop());
+    assert(adaptor.getPaddingLeft() == adaptor.getPaddingRight());
```
Comment (svuckovicTT) [inline lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp:248]: Add an error message, similar to the style of the assert in line ~31:
```
assert(false && "No device found");
```

File: lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp [hunk 23]
```diff
@@ -227,9 +288,11 @@ void populateTTIRToTTNNPatterns(MLIRContext *ctx, RewritePatternSet &patterns,
            EmbeddingOpConversionPattern,
            SoftmaxOpConversionPattern,
            TransposeOpConversionPattern,
+           SoftmaxOpConversionPattern,
```
Comment (svuckovicTT) [inline lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp:291]: Remove - the pattern is included 2 lines above.

File: lib/Dialect/TTIR/IR/TTIROps.cpp [hunk 24]
```diff
@@ -291,6 +291,25 @@ ::mlir::LogicalResult mlir::tt::ttir::MatmulOp::verify() {
...
+  ::mlir::RankedTensorType weightType = getWeight().getType();
+
+  if (inputType.getRank() < 3) {
+    return emitOpError("Input must be at least a 3D tensor");
+  }
+  if (weightType.getRank() != 4) {
+    return emitOpError("Weight must be a 4D tensor");
+  }
+  if (!getODSOperands(2).empty()) {
+    ::mlir::RankedTensorType biasType = getBias().getType();
```
Comment (svuckovicTT) [inline lib/Dialect/TTIR/IR/TTIROps.cpp:305]: There should be a `getBias()` method or similar in `TTIROps.h.inc`. Here's an example for an optional dtype attr I have locally for an op I added.
```
  ::mlir::tt::DataTypeAttr getDtypeAttr() {
    auto attr = ::llvm::dyn_cast_or_null<::mlir::tt::DataTypeAttr>(getProperties().dtype);
    return attr;
  }
```

Can you check if that'd work here?

File: lib/Dialect/TTIR/Transforms/Passes.cpp [hunk 25]
```diff
@@ -573,6 +573,11 @@ class TTIRLayoutDPSOperandsRewriter
     for (auto &operand : op->getOpOperands()) {
       bool isResult = op.isDpsInit(&operand);
 
+      // TTNN Conv2d moves input, weight, and bias from host to device
+      // itself. Inserting the ToLayoutOp on these operands is thus problematic.
+      if (mlir::isa<Conv2dOp>(op.getOperation()) && !isResult) {
+        continue;
+      }
```
Comment (svuckovicTT) [inline lib/Dialect/TTIR/Transforms/Passes.cpp:580]: It seems that there's some gotchas in how tensors are getting to device in general, thru C++ ttnn APIs... I'll chat with Artem to try and understand what the correct way would be to do these things, including the conv op. Can you point me to the code in ttnn I can reference, so that I can better explain what our needs are?

Comment (svuckovicTT) [reply lib/Dialect/TTIR/Transforms/Passes.cpp:580]: If we end up having to treat a couple ops like this, we should probably define a trait and tag those ops with it, should be cleaner that way ("data driven vs code driven").

File: runtime/include/tt/runtime/detail/ttnn.h [hunk 26]
```diff
@@ -59,6 +60,8 @@
 
 namespace tt::runtime::ttnn {
 
+constexpr std::size_t kL1SmallSize = 1 << 15;
```
Comment (svuckovicTT) [inline runtime/include/tt/runtime/detail/ttnn.h:63]: Can you add a comment explaining what this is for?

Comment (LPanosTT) [reply runtime/include/tt/runtime/detail/ttnn.h:63]: This is a temporary default L1Smallsize we will be using until we can allocate L1 per op more specifically. The original default is 0, which causes some buffers to not be created at all. I was running into issues with that while bringing up conv2d. @nsmithtt  anything to add?

File: CMakeLists.txt [hunk 27]
```diff
@@ -22,7 +22,7 @@ set(TTMLIR_ENABLE_BINDINGS_PYTHON ON CACHE BOOL "Enable Python bindings")
 
 list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/cmake/modules)
 
-add_compile_options(-Wall -Wextra -Wpedantic -Werror -Wno-unused-parameter --system-header-prefix=/opt/ttmlir-toolchain)
+add_compile_options(-Wall -Wextra -Wpedantic -Wno-unused-parameter --system-header-prefix=/opt/ttmlir-toolchain)
```
Comment (svuckovicTT) [inline CMakeLists.txt:25]: Not sure if accidental, but we shouldn't remove this flag.

Comment (LPanosTT) [reply CMakeLists.txt:25]: The uplifted version of TT-Metal has uint/int comparison warnings in it. Not sure what to do about that.

Comment (svuckovicTT) [reply CMakeLists.txt:25]: It's passing on main though? Or is it the conv file you included that's causing the issues?

Comment (LPanosTT) [reply CMakeLists.txt:25]: Its the conv file I've included yes

Comment (svuckovicTT) [reply CMakeLists.txt:25]: Gotta call in the big guns here... @nsmithtt please report for duty

Comment (nsmithtt) [reply CMakeLists.txt:25]: This cannot land, we definitely want `-Werror`.  Please add to the list of ignored warnings at the top of file `runtime/include/tt/runtime/detail/ttnn.h`. e.g.
```c++
#pragma clang diagnostic ignored "-Wundefined-inline"
```

File: include/ttmlir/Dialect/TTIR/IR/TTIROps.td [hunk 28]
```diff
@@ -369,6 +369,37 @@ def TTIR_ConcatOp : TTIR_DPSOp<"concat"> {
...
+
+    TT_OperandConstraintArrayAttr:$operand_constraints);
+
+    let results = (outs AnyRankedTensor:$result);
+
+    let extraClassDeclaration = [{
+      MutableOperandRange getDpsInitsMutable() { return getOutputMutable(); }
+    }];
+
+    let hasVerifier = 1;
```
[the remark this answers is not in the document]
Comment (nsmithtt) [reply include/ttmlir/Dialect/TTIR/IR/TTIROps.td:382]: Yeah, Lewis, I think you changed the whitespace of the op above Conv2d which makes the diff look like this.

File: runtime/include/tt/runtime/detail/ttnn.h (hunk 26 above)
Comment (nsmithtt) [reply runtime/include/tt/runtime/detail/ttnn.h:63]: It's a special allocator flag in TTNN that *must* be programmed during device creation when using convs.  Conv op specifically uses this region to store sideband indexing info for the op, it is kind of a special case in this regard.

File: lib/Dialect/TTIR/IR/TTIROps.cpp (hunk 24 above)
Comment (LPanosTT) [reply lib/Dialect/TTIR/IR/TTIROps.cpp]: It did

File: lib/Dialect/TTIR/Transforms/Passes.cpp (hunk 25 above)
Comment (LPanosTT) [reply lib/Dialect/TTIR/Transforms/Passes.cpp:580]: Ideally we'd want the conv op to work if all its inputs are on device already. For conv ops in the middle of a model it would be wasteful to have to move the inputs onto cpu just so the next conv can take them off.
